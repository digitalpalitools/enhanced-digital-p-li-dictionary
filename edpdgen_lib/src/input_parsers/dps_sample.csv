"Pāli1"	"Fin"	"POS"	"Grammar"	"Derived from"	"Neg"	"Verb"	"Trans"	"Case"	"Meaning IN CONTEXT"	"Meaning in native language"	"Pāli Root"	"Base"	"Construction"	"Sanskrit"	"Sk Root"	"Commentary"	"Notes"	"Source1"	"Example1"	"Sutta1"	"Source 2"	"Example 2"	"Sutta2"	"Chapter"	"Test"
"kaṇṇa 1"	"ps"	"nt"	""	""	""	""	""	""	"ear"	"ухо"	""	""	""	"karṇa"	""	""	""	"AN 7.61"	"tato tvaṃ, moggallāna, ubho <b>kaṇṇa</b>-sotāni āviñcheyyāsi, pāṇinā gattāni anumajjeyyāsi."	"pacalāyamāna sutta"	"MN 51"	"pharusaṃ vācaṃ pahāya pharusāya vācāya paṭivirato hoti, yā sā vācā n’elā <b>kaṇṇa</b>-sukhā pemanīyā hadayaṅ’gamā porī bahujanakantā bahujanamanāpā tathārūpiṃ vācaṃ bhāsitā hoti"	"anupubba-sikkhā"	"Teachings"	"5"
"jetavana"	"ps"	"m"	"comp"	""	""	""	""	""	"Jeta’s Grove"	"роща Джеты"	""	""	"jetassa + vana"	"jetavana"	""	""	""	"SN 22.33"	"seyyathā’pi, bhikkhave, yaṃ imasmiṃ <b>jetavane</b> tiṇakaṭṭhasākhāpalāsaṃ taṃ jano hareyya vā ḍaheyya vā yathāpaccayaṃ vā kareyya."	"natumhākaṃ sutta"	"SNP 16 mahāvaggo 5"	"evaṃ me sutaṃ — ekaṃ samayaṃ bhagavā sāvatthiyaṃ viharati <b>jetavane</b> anāthapiṇḍikassa ārāme."	"maṅgala-sutta"	"Paritta"	"5"
"dasama"	"p"	"ordin"	""	"dasa"	""	""	""	""	"tenth (10th)"	"десятый (10-й)"	""	""	"dasa + ma"	"daśama"	""	""	""	"Sri Lanka"	"aṭṭhamaṃ adhiṭṭhāna-pārami, navamaṃ mettā-pārami, <b>dasamaṃ</b> upekkhā-pāramī'ti"	"buddha-vassa-sajjhāyanā"	""	""	""	""	"5"
"paṭipucchā"	"p"	"f"	""	"pucchā"	""	""	""	""	"a question in return"	"ответный вопрос, вопрос в ответ"	"√pucch 1.1 a (to ask, question)"	""	"pati + √pucch + ā"	""	"√prach "	""	""	"AN 4.42"	"atthi bhikkhave pañho ek’aṃsa-byākaraṇīyo, atthi pañho vibhajja-byākaraṇīyo, atthi pañho <b>paṭipucchā</b>-byākaraṇīyo; atthi pañho ṭhapanīyo."	"pañhabyākaraṇa sutta"	""	""	""	""	"5"
"samañchati"	"p"	"pr"	""	""	""	""	"trans"	"+acc"	"stretches out, pulls out, drags out"	"тянет, растягивает, вытягивает"	"√añch 1.1 a (to extend, draw along)"	"√añch + a > añcha"	"saṃ + añcha + ti"	"saṃ + āñchati"	"√āñch"	""	""	"VIN 2.5.9.7"	"kissa tvaṃ, udāyi, nisīdanaṃ samantato <b>samañchasi</b>, seyyathā'pi purāṇ’āsikoṭṭho'ti?"	"nisīdana sikkhāpadaṃ"	""	""	""	""	"5"
