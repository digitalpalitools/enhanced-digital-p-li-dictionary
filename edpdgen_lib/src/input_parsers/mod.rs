use crate::EdpdLogger;
use regex::{Captures, Regex};
use std::path::Path;
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "toc_summary",
            include_str!("templates/toc_summary.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.add_raw_templates(vec![(
            "word_data",
            include_str!("templates/word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
    static ref PALI1_CRACKER: Regex = Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

pub trait PaliWord {
    fn sort_key(&self) -> String;
    fn group_id(&self) -> String;
    fn toc_id(&self) -> String;
    fn toc_entry(&self) -> Result<String, String>;
    fn word_data_entry(&self, short_name: &str) -> Result<String, String>;
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DpdPaliWord {
    #[serde(rename = "Pāli1")]
    pali1: String,
    #[serde(rename = "Pāli2")]
    pali2: String,
    #[serde(rename = "Fin")]
    fin: String,
    #[serde(rename = "POS")]
    pos: String,
    #[serde(rename = "Grammar")]
    grammar: String,
    #[serde(rename = "Derived from")]
    derived_from: String,
    #[serde(rename = "Neg")]
    neg: String,
    #[serde(rename = "Verb")]
    verb: String,
    #[serde(rename = "Trans")]
    trans: String,
    #[serde(rename = "Case")]
    case: String,
    #[serde(rename = "Meaning IN CONTEXT")]
    in_english: String,
    #[serde(rename = "Sanskrit")]
    sanskrit: String,
    #[serde(rename = "Sk Root")]
    sanskrit_root: String,
    #[serde(rename = "Family")]
    family: String,
    #[serde(rename = "Pāli Root")]
    pali_root: String,
    #[serde(rename = "V")]
    v: String,
    #[serde(rename = "Grp")]
    grp: String,
    #[serde(rename = "Sgn")]
    sgn: String,
    #[serde(rename = "Root Meaning")]
    root_meaning: String,
    #[serde(rename = "Base")]
    base: String,
    #[serde(rename = "Construction")]
    construction: String,
    #[serde(rename = "Derivative")]
    derivative: String,
    #[serde(rename = "Suffix")]
    suffix: String,
    #[serde(rename = "Compound")]
    compound: String,
    #[serde(rename = "Compound Construction")]
    compound_construction: String,
    #[serde(rename = "Source1")]
    source1: String,
    #[serde(rename = "Sutta1")]
    sutta1: String,
    #[serde(rename = "Example1")]
    example1: String,
    #[serde(rename = "Source 2")]
    source2: String,
    #[serde(rename = "Sutta2")]
    sutta2: String,
    #[serde(rename = "Example 2")]
    example2: String,
    #[serde(rename = "Antonyms")]
    antonyms: String,
    #[serde(rename = "Synonyms – different word")]
    synonyms: String,
    #[serde(rename = "Variant – same constr or diff reading")]
    variant: String,
    #[serde(rename = "Commentary")]
    commentary: String,
    #[serde(rename = "Notes")]
    notes: String,
    #[serde(rename = "Stem")]
    stem: String,
    #[serde(rename = "Pattern")]
    pattern: String,
    #[serde(rename = "Buddhadatta")]
    buddhadatta: String,
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpdPaliWord,
    toc_id: &'a str,
    short_name: &'a str,
}

impl PaliWord for DpdPaliWord {
    fn sort_key(&self) -> String {
        let sk = PALI1_CRACKER.replace(&self.pali1, |caps: &Captures| {
            // NOTE: Best case effort. Not sweating it.
            let n = &caps[3].parse::<i32>().unwrap_or(0);
            format!("{} {:03}", &caps[1], n)
        });

        sk.into_owned()
    }

    fn group_id(&self) -> String {
        let gid = PALI1_CRACKER.replace(&self.pali1, |caps: &Captures| caps[1].to_string());

        gid.into_owned()
    }

    fn toc_id(&self) -> String {
        self.pali1.replace(" ", "_") + "_dpd"
    }

    fn toc_entry(&self) -> Result<String, String> {
        let mut context = Context::new();
        context.insert("toc_id", &self.toc_id());
        context.insert("pali1", &self.pali1);
        context.insert("pos", &self.pos);
        context.insert("in_english", &self.in_english);

        TEMPLATES
            .render("toc_summary", &context)
            .map_err(|e| e.to_string())
    }

    fn word_data_entry(&self, short_name: &str) -> Result<String, String> {
        let vm = WordDataViewModel {
            word: &self,
            toc_id: &self.toc_id(),
            short_name,
        };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("word_data", &context)
            .map_err(|e| e.to_string())
    }
}

pub fn load_words<'a>(
    path: &Path,
    logger: &'a impl EdpdLogger,
) -> Result<impl Iterator<Item = impl PaliWord> + 'a, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    let words = rdr
        .into_deserialize::<DpdPaliWord>()
        .enumerate()
        .filter_map(move |(i, r)| match r {
            Ok(w) => Some(w),
            Err(e) => {
                logger.error(&format!(
                    "Unable to deserialize record #{}. Error: {}.",
                    i, e
                ));
                None
            }
        });

    logger.info(&format!("... done loading words from {:?}.", &path));
    Ok(words)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0, "ābādha"; "0 digits")]
    #[test_case(9, "adhikāra 001"; "1 digit")]
    #[test_case(11, "adhikāra 010"; "2 digits")]
    fn test_sort_key(rec_number: usize, expected_sk: &str) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("failed to load");

        let sk = recs
            .nth(rec_number)
            .map(|r| r.sort_key())
            .expect("unexpected");

        assert_eq!(sk, expected_sk)
    }

    #[test_case(4, "abahulīkata"; "0 digits")]
    #[test_case(5, "abala"; "1 digit")]
    #[test_case(11, "adhikāra"; "2 digits")]
    fn test_group_id(rec_number: usize, expected_gid: &str) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("failed to load");

        let gid = recs
            .nth(rec_number)
            .map(|r| r.group_id())
            .expect("unexpected");

        assert_eq!(gid, expected_gid);
    }

    #[test_case(4, "abahulīkata_dpd"; "0 digits")]
    #[test_case(5, "abala_1_dpd"; "1 digit")]
    #[test_case(11, "adhikāra_10_dpd"; "2 digits")]
    fn test_toc_id(rec_number: usize, expected_toc_id: &str) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("unexpected");

        let toc_id = recs
            .nth(rec_number)
            .map(|r| r.toc_id())
            .expect("unexpected");

        assert_eq!(toc_id, expected_toc_id);
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn toc_summary_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("unexpected");

        let toc_summary = recs
            .nth(rec_number)
            .map(|r| r.toc_entry().expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(toc_summary);
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("unexpected");

        let word_data = recs
            .nth(rec_number)
            .map(|r| r.word_data_entry("dpd").expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
