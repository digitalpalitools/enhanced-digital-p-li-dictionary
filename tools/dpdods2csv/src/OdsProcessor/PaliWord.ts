export interface PaliWordBase {
  sortKey(): string
  groupId(): string
  tocId(): string
  includeInRootCsv(): boolean
  includeInDictionary(): boolean
  toCsvRow(): string
  createTocSummary(): string
  createWordData(): string
}

export type PaliWordFactory = (x: string[]) => PaliWordBase

export const padTrailingNumbers = (s: string) => s.replace(/\d+/g, (m) => '00'.substr(m.length - 1) + m)

export const makeGroupId = (baseGroupId: string) => {
  const parts = baseGroupId.split(' ')

  if (!Number.isNaN(parseInt(parts[parts.length - 1], 10))) {
    return parts.slice(0, parts.length - 1).join(' ')
  }

  return baseGroupId
}

export const toCsvRow = (records: string[]) => records.map((x) => `"${x}"`).join('\t')
