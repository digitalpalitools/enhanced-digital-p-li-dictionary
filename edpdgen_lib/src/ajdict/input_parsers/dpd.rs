use crate::ajdict::AjDictPaliWord;
use crate::input::dpd::DpdPaliWord;
use crate::input::make_sort_key;
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "dpd_word_data",
            include_str!("templates/dpd_word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera
    };
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpdPaliWord,
}

impl AjDictPaliWord for DpdPaliWord {
    fn id(&self) -> &str {
        &self.pali1
    }

    fn sort_key(&self) -> String {
        make_sort_key(&self.id())
    }

    fn word_data_entry(&self) -> Result<String, String> {
        let vm = WordDataViewModel { word: &self };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("dpd_word_data", &context)
            .map(|x| x.replace("\r", "").replace("\n", ""))
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ajdict::input_parsers::load_words;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let word_data = recs
            .nth(rec_number)
            .map(|r| r.word_data_entry().expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
