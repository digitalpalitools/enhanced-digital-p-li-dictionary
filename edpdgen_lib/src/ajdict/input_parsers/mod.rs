use crate::ajdict::AjDictPaliWord;
use crate::EdpdLogger;
use regex::{Captures, Regex};
use std::path::Path;

pub mod dpd;

lazy_static! {
    static ref PALI1_CRACKER: Regex = Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

//
//
//
//
//  DUP code
//
//
//
//
//
//
pub fn load_words<'a, T: 'a + serde::de::DeserializeOwned + AjDictPaliWord>(
    path: &Path,
    logger: &'a dyn EdpdLogger,
) -> Result<impl Iterator<Item = impl AjDictPaliWord> + 'a, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    let words = rdr
        .into_deserialize::<T>()
        .enumerate()
        .filter_map(move |(i, r)| match r {
            Ok(w) => Some(w),
            Err(e) => {
                logger.error(&format!(
                    "Unable to deserialize record #{}. Error: {}.",
                    i, e
                ));
                None
            }
        });

    logger.info(&format!("... done loading words from {:?}.", &path));
    Ok(words)
}

//
//
//
//
//  DUP code
//
//
//
//
//
//
fn make_sort_key(id: &str) -> String {
    let sk = PALI1_CRACKER.replace(id, |caps: &Captures| {
        // NOTE: Best case effort. Not sweating it.
        let n = &caps[3].parse::<i32>().unwrap_or(0);
        format!("{} {:03}", &caps[1], n)
    });

    sk.into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::input::dpd::DpdPaliWord;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0, "훮b훮dha"; "0 digits")]
    #[test_case(9, "adhik훮ra 001"; "1 digit")]
    #[test_case(11, "adhik훮ra 010"; "2 digits")]
    fn test_sort_key(rec_number: usize, expected_sk: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let sk = recs
            .nth(rec_number)
            .map(|r| make_sort_key(r.id()))
            .expect("unexpected");

        assert_eq!(sk, expected_sk)
    }
}
