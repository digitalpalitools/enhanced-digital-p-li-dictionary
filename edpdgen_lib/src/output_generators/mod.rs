use crate::input_parsers::PaliWord;
use crate::EdpdLogger;
use itertools::Itertools;
use std::cmp::Ordering;

// use std::fs::File;
// use std::io::Write;

use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "word_group",
            include_str!("templates/word_group.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
}

#[derive(Debug, Serialize)]
struct IdxEntry {
    word: String,
    data_offset: i32,
    data_size: i32,
}

/**
 * From: https://stackoverflow.com/a/13225961/6196679
 *
 * Compares two strings, ignoring the case of ASCII characters. It treats
 * non-ASCII characters taking in account case differences. This is an
 * attempt to mimic glib's string utility function
 * <a href="http://developer.gnome.org/glib/2.28/glib-String-Utility-Functions.html#g-ascii-strcasecmp">g_ascii_strcasecmp ()</a>.
 *
 * This is a slightly modified version of java.lang.String.CASE_INSENSITIVE_ORDER.compare(String s1, String s2) method.
 *
 * @param str1  string to compare with str2
 * @param str2  string to compare with str1
 * @return      0 if the strings match, a negative value if str1 < str2, or a positive value if str1 > str2
 */
pub fn g_ascii_strcasecmp(str1: &str, str2: &str) -> Ordering {
    let str_vec1: Vec<char> = str1.chars().collect();
    let str_vec2: Vec<char> = str2.chars().collect();
    let n1 = str_vec1.len();
    let n2 = str_vec2.len();
    let c127 = 127 as char;

    let min = n1.min(n2);
    for i in 0..min {
        let c1 = str_vec1[i];
        let c2 = str_vec2[i];
        if c1 != c2 {
            if c1 > c127 || c2 > c127 {
                // If non-ASCII char...
                return c1.cmp(&c2);
            } else {
                let c1uc = c1.to_ascii_uppercase();
                let c2uc = c2.to_ascii_uppercase();
                if c1uc != c2uc {
                    let c1lc = c1.to_ascii_lowercase();
                    let c2lc = c2.to_ascii_lowercase();
                    if c1lc != c2lc {
                        return c1lc.cmp(&c2lc);
                    }
                }
            }
        }
    }

    n1.cmp(&n2)
}

#[derive(Serialize)]
struct WordGroupViewModel<'a> {
    ods_type: &'a str,
    accent_color: &'a str,
    toc_entries: &'a [String],
    descriptions: &'a [String],
}

fn create_html_for_word_group(
    ods_type: &str,
    words: impl Iterator<Item = impl PaliWord>,
    _logger: &impl EdpdLogger,
) -> Result<String, String> {
    let mut word_info: Vec<(String, String, String)> = words
        .map(|w| {
            (
                w.sort_key(),
                w.toc_entry().unwrap_or_else(|e| e),
                w.word_data_entry().unwrap_or_else(|e| e),
            )
        })
        .collect();
    word_info.sort_by(|a, b| a.0.cmp(&b.0));

    let toc_entries: Vec<String> = word_info.iter().map(|w| w.1.to_owned()).collect();
    let descriptions: Vec<String> = word_info.iter().map(|w| w.2.to_owned()).collect();

    let vm = WordGroupViewModel {
        ods_type,
        accent_color: "orange",
        toc_entries: &toc_entries,
        descriptions: &descriptions,
    };

    let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
    TEMPLATES
        .render("word_group", &context)
        .map_err(|e| e.to_string())
}

fn create_dict(
    ods_type: &str,
    words: impl Iterator<Item = impl PaliWord>,
    logger: &impl EdpdLogger,
) -> Result<(Vec<u8>, Vec<IdxEntry>), String> {
    let word_groups = words.group_by(|pw| pw.group_id());

    let mut dict_buffer: Vec<u8> = Vec::new();
    let mut idx_words: Vec<IdxEntry> = Vec::new();
    for (n, (key, word_group)) in (&word_groups).into_iter().enumerate() {
        let html_str = create_html_for_word_group(ods_type, word_group, logger)?;
        let mut html_bytes = html_str.into_bytes();

        // //if key.eq("ajapada") {//"abbahe") {
        // logger.info(&format!("Writing {}...", key));
        // let mut dict_file = File::create(format!("d:/delme/dicts-rust/words/{}.txt", key)).map_err(|e| e.to_string()).unwrap();
        // dict_file.write_all(&html_bytes).unwrap();
        // //}

        idx_words.push(IdxEntry {
            word: key,
            data_offset: if n == 0 {
                0
            } else {
                idx_words[n - 1].data_offset + idx_words[n - 1].data_size
            },
            data_size: html_bytes.len() as i32,
        });
        dict_buffer.append(&mut html_bytes);
    }

    Ok((dict_buffer, idx_words))
}

fn create_idx(idx_entries: &mut Vec<IdxEntry>, logger: &impl EdpdLogger) -> Vec<u8> {
    idx_entries.sort_by(|w1, w2| g_ascii_strcasecmp(&w1.word, &w2.word));

    let idx: Vec<u8> = idx_entries.iter().fold(Vec::new(), |mut acc, e| {
        acc.append(&mut e.word.to_owned().into_bytes());
        acc.push(0u8);
        acc.extend_from_slice(&e.data_offset.to_be_bytes());
        acc.extend_from_slice(&e.data_size.to_be_bytes());
        acc
    });

    logger.info(&format!("Created {} idx entries.", &idx_entries.len()));
    idx
}

pub fn write_dictionary(
    ods_type: &str,
    words: impl Iterator<Item = impl PaliWord>,
    logger: &impl EdpdLogger,
) -> Result<(Vec<u8>, Vec<u8>), String> {
    let (dict, mut idx_entries) = create_dict(ods_type, words, logger)?;

    let idx = create_idx(&mut idx_entries, logger);

    Ok((dict, idx))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;

    #[derive(Debug, Deserialize)]
    struct TestPaliWord {
        sort_key: String,
        group_id: String,
        toc_id: String,
        toc_entry: String,
        word_data_entry: String,
    }

    impl PaliWord for TestPaliWord {
        fn sort_key(&self) -> String {
            self.sort_key.clone()
        }

        fn group_id(&self) -> String {
            self.group_id.clone()
        }

        fn toc_id(&self) -> String {
            self.toc_id.clone()
        }

        fn toc_entry(&self) -> Result<String, String> {
            Ok(self.toc_entry.clone())
        }

        fn word_data_entry(&self) -> Result<String, String> {
            Ok(self.word_data_entry.clone())
        }
    }

    fn read_pali_words<'a>() -> impl Iterator<Item = impl PaliWord> + 'a {
        let path = resolve_file_in_manifest_dir("src/output_generators/test_data/pali_words1.csv")
            .expect("must exist!");

        let file = std::fs::File::open(path).expect("must exist");
        let rdr = csv::ReaderBuilder::new().from_reader(file);

        rdr.into_deserialize::<TestPaliWord>().map(|w| w.expect(""))
    }

    #[test]
    fn create_dict_test() {
        let words = read_pali_words();

        let (dict_data, idx_entries) =
            create_dict("dpd", words, &TestLogger::new()).expect("Unexpected");
        let dict_entries: Vec<String> = idx_entries
            .iter()
            .map(|ie| {
                let word_bytes =
                    &dict_data[ie.data_offset as usize..(ie.data_offset + ie.data_size) as usize];
                std::str::from_utf8(word_bytes)
                    .expect("unexpected")
                    .to_owned()
            })
            .collect();

        insta::assert_yaml_snapshot!(idx_entries);
        insta::assert_debug_snapshot!(dict_entries);
    }

    #[test]
    fn create_idx_test() {
        let mut idx_entries = vec![
            IdxEntry {
                word: "a".to_string(),
                data_offset: 1,
                data_size: 2,
            },
            IdxEntry {
                word: "bc".to_string(),
                data_offset: 0x01000100,
                data_size: 0x00020002,
            },
        ];

        let idx = create_idx(&mut idx_entries, &TestLogger::new());

        assert_eq!(
            idx,
            vec![0x61, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0x62, 0x63, 0, 1, 0, 1, 0, 0, 2, 0, 2]
        )
    }
}
