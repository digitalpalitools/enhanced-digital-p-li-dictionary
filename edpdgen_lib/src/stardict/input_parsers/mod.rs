use crate::inflection_generator::InflectionGenerator;
use crate::{EdpdLogger, InputFormat};
use regex::{Captures, Regex};
use std::path::Path;

pub mod dpd;
pub mod dps;

lazy_static! {
    static ref PALI1_CRACKER: Regex = Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

pub trait PaliWord {
    fn id(&self) -> &str;
    fn sort_key(&self) -> String;
    fn group_id(&self) -> String;
    fn toc_id(&self, input_format: &InputFormat) -> String;
    fn toc_entry(&self, input_format: &InputFormat) -> Result<String, String>;
    fn word_data_entry(
        &self,
        input_format: &InputFormat,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String>;
}

pub fn load_words<'a, T: 'a + serde::de::DeserializeOwned + PaliWord>(
    path: &Path,
    logger: &'a dyn EdpdLogger,
) -> Result<impl Iterator<Item = impl PaliWord> + 'a, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    let words = rdr
        .into_deserialize::<T>()
        .enumerate()
        .filter_map(move |(i, r)| match r {
            Ok(w) => Some(w),
            Err(e) => {
                logger.error(&format!(
                    "Unable to deserialize record #{}. Error: {}.",
                    i, e
                ));
                None
            }
        });

    logger.info(&format!("... done loading words from {:?}.", &path));
    Ok(words)
}

fn make_sort_key(id: &str) -> String {
    let sk = PALI1_CRACKER.replace(id, |caps: &Captures| {
        // NOTE: Best case effort. Not sweating it.
        let n = &caps[3].parse::<i32>().unwrap_or(0);
        format!("{} {:03}", &caps[1], n)
    });

    sk.into_owned()
}

fn make_group_id(id: &str) -> String {
    let gid = PALI1_CRACKER.replace(id, |caps: &Captures| caps[1].to_string());

    gid.into_owned()
}

fn make_toc_id(id: &str, input_format: &InputFormat) -> String {
    format!("{}_{}", id.replace(" ", "_"), input_format.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resolve_file_in_manifest_dir;
    use crate::stardict::input_parsers::dpd::DpdPaliWord;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0, "ābādha"; "0 digits")]
    #[test_case(9, "adhikāra 001"; "1 digit")]
    #[test_case(11, "adhikāra 010"; "2 digits")]
    fn test_sort_key(rec_number: usize, expected_sk: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let sk = recs
            .nth(rec_number)
            .map(|r| make_sort_key(r.id()))
            .expect("unexpected");

        assert_eq!(sk, expected_sk)
    }

    #[test_case(4, "abahulīkata"; "0 digits")]
    #[test_case(5, "abala"; "1 digit")]
    #[test_case(11, "adhikāra"; "2 digits")]
    fn test_group_id(rec_number: usize, expected_gid: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let gid = recs
            .nth(rec_number)
            .map(|r| make_group_id(r.id()))
            .expect("unexpected");

        assert_eq!(gid, expected_gid);
    }

    #[test_case(4, "abahulīkata_dps"; "0 digits")]
    #[test_case(5, "abala_1_dps"; "1 digit")]
    #[test_case(11, "adhikāra_10_dps"; "2 digits")]
    fn test_toc_id(rec_number: usize, expected_toc_id: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let toc_id = recs
            .nth(rec_number)
            .map(|r| make_toc_id(r.id(), &InputFormat::DevamittaPaliStudy))
            .expect("unexpected");

        assert_eq!(toc_id, expected_toc_id);
    }
}
