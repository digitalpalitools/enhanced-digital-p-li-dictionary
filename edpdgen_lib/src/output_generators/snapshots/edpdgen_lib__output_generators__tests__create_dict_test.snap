---
source: edpdgen_lib/src/output_generators/mod.rs
expression: idx_entries

---
- word: ābādha
  data_offset: 0
  data_size: 313
- word: ābādhato
  data_offset: 313
  data_size: 315
- word: ābādhika
  data_offset: 628
  data_size: 315
- word: ābādhikinī
  data_offset: 943
  data_size: 318
- word: abahulīkata
  data_offset: 1261
  data_size: 317
- word: abala
  data_offset: 1578
  data_size: 376
- word: abalabala
  data_offset: 1954
  data_size: 314
- word: ābandhaka
  data_offset: 2268
  data_size: 315
- word: adhikāra
  data_offset: 2583
  data_size: 492

